druid/src/widget/viewport.rs

```diff
@@ -0,0 +1,259 @@
+// Copyright 2020 The Druid Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::kurbo::{Affine, Point, Rect, Size, Vec2};
+use crate::widget::prelude::*;
+use crate::{Data, WidgetPod};
+
+/// Represents the size and position of a rectangular "viewport" into a larger area.
+#[derive(Clone, Copy, Default, Debug, PartialEq)]
+pub struct RectPort {
+    /// The size of the area that we have a viewport into.
+    pub content_size: Size,
+    /// The view rectangle. This should be contained in `content_size.to_rect()`.
+    pub rect: Rect,
+}
+
+impl RectPort {
+    /// Checks whether `origin` is a valid origin for the viewport rectangle; if it isn't valid,
+    /// tries to move it to the nearest valid origin. (Here, an origin is valid if the viewport,
+    /// when translated to that origin, fits entirely within the content rectangle.)
+    pub fn clamp_view_origin(&self, origin: Point) -> Point {
+        let x = origin
+            .x
+            .min(self.content_size.width - self.rect.width())
+            .max(0.0);
+        let y = origin
+            .y
+            .min(self.content_size.height - self.rect.height())
+            .max(0.0);
+        Point::new(x, y)
+    }
+
+    /// Changes the viewport offset by `delta`.
+    ///
+    /// Returns true if the offset actually changed. Even if `delta` is non-zero, the offset might
+    /// not change. For example, if you try to move the viewport down but it is already at the
+    /// bottom of the child widget, then the offset will not change and this function will return
+    /// false.
+    pub fn scroll_by(&mut self, delta: Vec2) -> bool {
+        self.scroll_to(self.rect.origin() + delta)
+    }
+
+    /// Sets the viewport origin to `pos`.
+    ///
+    /// Returns true if the position changed. Note that the valid values for the viewport origin
+    /// are constrained by the size of the child, and so the origin might not get set to exactly
+    /// `pos`.
+    pub fn scroll_to(&mut self, origin: Point) -> bool {
```

Comment:
Similar to above, perhaps `offset_to` or `pan_to`?

Reply:
Snap!